crates/cow-amm/src/implementations/safe_based/cow_amm_constant_product_factory.rs

```diff
@@ -0,0 +1,58 @@
+use {
+    crate::{implementations::safe_based::cow_amm::CowAmm, indexer::CowAmmRegistry},
+    ethcontract::{common::DeploymentInformation, dyns::DynWeb3},
+    shared::impl_event_retrieving,
+    std::sync::Arc,
+};
+
+impl_event_retrieving! {
+    pub Contract for contracts::cow_amm_constant_product_factory
+}
+
+pub struct CowAmmConstantProductFactoryHandler {
```

Comment:
I think it should be possible to get rid of this intermediate type.
We still need the trait that turns a given event into something like an `Option<impl CowAmm>` but this trait can be implemented for the event enums themselves.
Then we only need to figure out now to get rid of the `deployment_info` stuff. 🤔